Add basic custom metrics support

zhangtbj opened the pull request:
Based in our controller implementations, I add 6 basic custom metrics:
- controller up status
- build count
- buildrun count
- buildrun establish time
- buildrun execution time
- buildrun execution time histogram

I also verified after the controller is retarted and the metrics is cleaned up, the later metrics can still show correctly.

I also add ut for metrics.

Here is the example metrics show from Sysdig:
<img width="1349" alt="截屏2020-05-29 上午11 59 30" src="https://user-images.githubusercontent.com/18108042/83219831-f9264f00-a1a3-11ea-8f34-57303815f1cf.png">
<img width="1349" alt="截屏2020-05-29 上午11 59 46" src="https://user-images.githubusercontent.com/18108042/83219839-fcb9d600-a1a3-11ea-9440-13971f7c6b63.png">

otaviof commented:
Thanks! That looks really good :-) I really like the graphs, I think you've crafted a well placed change to add those metrics. Kudos!

otaviof commented:
By the way, would be good to add specific documentation about the metrics the operator exposes. In other projects we usually see the description of each metric exported.

zhangtbj commented:
Sure thing, let me add a document in the PR :)

zhangtbj commented:
Hi @otaviof ,

I also added a `metrics.md` in my PR, it is:
https://github.com/zhangtbj/build/blob/custom_metric/docs/metrics.md

And provide another 6 hours charts :)
<img width="1337" alt="截屏2020-05-29 下午6 23 43" src="https://user-images.githubusercontent.com/18108042/83249712-9354ba00-a1d9-11ea-900a-8c17f88516b7.png">

zhangtbj commented:
/retest

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+var (
```
This is confusing. I will keep it clean and only define the variable types. Also I dont think this is the place to do this because this class should only define what can you do with the `metrics` package, not how to call it.
It should be:
```
var (
	upStatus                                                                    prometheus.Gauge
	buildCount, buildRunCount, buildRunEstablishDuration, buildRunReadyDuration prometheus.GaugeVec
	buildRunReadyDurationHis                                                    prometheus.HistogramVec
)
```

and should be defined somewhere else not here. Wherever you wanna call the pkg.

zhangtbj replied:
I suggest we keep this style first until we finalize them in future phase 2.

For example, `buildCount`, `buildRunCount` may change to `COUNT` type just increase the count.

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+//ProxyMetrics of the package level item.
+var BuildMetrics = CustomBuildMetrics{
```
This is not a good approach. You hardcode a variable inside this package with a predefined value. This class should only define a function on how to get an instance of `CustomBuildMetrics`, like:

```
// NewCustomBuildMetrics ...
func NewCustomBuildMetrics(
	upStatus prometheus.Gauge,
	buildC *prometheus.GaugeVec,
	buildRunC *prometheus.GaugeVec,
	buildRunED *prometheus.GaugeVec,
	buildRunRD *prometheus.GaugeVec,
	buildRunRDH *prometheus.HistogramVec,
) *CustomBuildMetrics {
	return &CustomBuildMetrics{
		UpStatus:                  upStatus,
		BuildCount:                buildC,
		BuildRunCount:             buildRunC,
		BuildRunEstablishDuration: buildRunED,
		BuildRunReadyDuration:     buildRunRD,
		BuildRunReadyDurationHis:  buildRunRDH,
	}
}
```
Please modify this.

zhangtbj replied:
Yep, also comes from an example struct, improved to use the simple struct

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+//Describe Implementation of the Collector Describe interface function
+func (r CustomBuildMetrics) Describe(descCh chan<- *prometheus.Desc) {
```
for what is this func? and who uses it?

zhangtbj replied:
It should be used in future for some additional metric operations, I copied from example. But OK let me remove it first.

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+//Collect Implementation of the Collector Collect interface function
+func (r CustomBuildMetrics) Collect(collectCh chan<- prometheus.Metric) {
```
for what is this func? and who uses it?

zhangtbj replied:
It should be used in future for some additional metric operations, I copied from example. But OK let me remove it first.

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+	BuildRunReadyDurationHis:         *buildRunReadyDurationHis,
+}
+
+//Describe Implementation of the Collector Describe interface function
```
leave empty space between `//` and the comment.

zhangtbj replied:
removed this func

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+	r.BuildRunReadyDurationHis.Describe(descCh)
+}
+
+//Collect Implementation of the Collector Collect interface function
```
leave empty space between `//` and the comment.

zhangtbj replied:
removed this func

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+	BuildRunReadyDurationHis      prometheus.HistogramVec
+}
+
+//ProxyMetrics of the package level item.
```
leave empty space between `//` and the comment.

zhangtbj replied:
fixed

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+//If the build-controller service is up and running, the UpStatus should always return 1
+func init() {
```
Some comments on this:
1. I dont think you should call this here. If I get this correctly, this is to start the metrics mechanism, so it should be call on a different package. From the fact that you are using init(), I will say this should go inmediately after controllers are started.
2. You should use the function I mentioned above to get an instance of `CustomBuildMetrics`, so this init func should look like:
```
//If the build-controller service is up and running, the UpStatus should always return 1
func init() {

	us := prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "up_status",
		Help: "The up status of the build-controller.",
	})
	bc := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "build_count",
			Help: "Build creation count.",
		},
		[]string{"build"})
	brc := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "buildrun_count",
			Help: "BuildRun execution count.",
		},
		[]string{"buildrun"})

	bred := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "buildrun_establish_duration",
			Help: "BuildRun establish duration in seconds.",
		},
		[]string{"buildrun"})

	brrd := prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Name: "buildrun_ready_duration",
			Help: "BuildRun ready duration in seconds.",
		},
		[]string{"buildrun"})

	brrdh := prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "buildrun_ready_duration_his",
			Help:    "BuildRun ready duration histogram in seconds.",
			Buckets: prometheus.LinearBuckets(50, 50, 10),
		},
		[]string{"buildrun"})

	bm := NewCustomBuildMetrics(us, bc, brc, bred, brrd, brrdh)
	bm.ControllerUp()

	// BuildMetrics.ControllerUp()
	// Register custom metrics with the global prometheus registry
	metrics.Registry.MustRegister(
		bm.UpStatus,
		bm.BuildCount,
		bm.BuildRunCount,
		bm.BuildRunEstablishDuration,
		bm.BuildRunReadyDuration,
		bm.BuildRunReadyDuration,
		bm.BuildRunReadyDurationHis)
}
```

zhangtbj replied:
No, they are the metric type initialization, it should be done at the beginning. init() func should to start and register metric. They are at different level.

And if I put them to the init() func, other initialization will report error

qu1queee replied:
Can we externalize this init into the main.go or similar, during controllers initialization. As I mentioned, the reason is that this package should define only what is needed for the metrics, not the call to get an instance.

qu1queee replied:
See this example from @SaschaSchwarze0 spike, https://github.com/SaschaSchwarze0/build/blob/985d2716a0f38ce83885751333acae9c5273bc2f/cmd/manager/main.go#L139-L141

gabemontero replied:
Yeah in OCP itself we've avoided metric initialization in `init()` functions as well, using patterns not too dissimilar from the one noted above.

@adambkaplan - I almost remember a best practice somewhere talking about it, but I can't find it now.  Maybe I am misremembering.  Do you recall?

Perhaps we should ping our old OCP monitoring/prometheus colleagues from Group D ?  

WDYT

sbose78 replied:
Could you please summarize/address this @zhangtbj ? 

@brancz from monitoring has already been on this PR :)

zhangtbj replied:
Hi @SaschaSchwarze0 ,

Did you have a try to init the metrics and check the metrics locally by using https://github.com/SaschaSchwarze0/build/blob/985d2716a0f38ce83885751333acae9c5273bc2f/cmd/manager/main.go#L139-L141 ?

I changed to this way, but it doesn't work and I cannot get the metrics correctly in local.

Or any idea or info about that @adambkaplan or @brancz

SaschaSchwarze0 replied:
Hi @zhangtbj, yes, the code in my spike branch used to work when running the controller threw `make local`. What misbehavior are you getting? Is there an exception? Are the metrics not registered?

zhangtbj replied:
Hi @SaschaSchwarze0 ,
Thanks for the info. Today I re-ran again, I can see the metric now.

So I switched to use:
- change `init()` to `func InitPrometheus(config *config.Config)`
- Add `buildMetrics.InitPrometheus(c)` in main.go

It works, and I updated in the commit.

FYI @sbose78 and @qu1queee

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+		buildRunReadyDurationHis, )
+}
+
+func (r CustomBuildMetrics) ControllerUp() {
```
pls add comments to the function

zhangtbj replied:
added comments

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+func (r CustomBuildMetrics) ControllerUp() {
+	r.UpStatus.Set(1)
```
what does `1` means?

zhangtbj replied:
1 means the controller pod starts up.

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+	r.UpStatus.Set(1)
+}
+
+func (r CustomBuildMetrics) BuildCountInc(buildLabel string) {
```
pls add comments to the function

zhangtbj replied:
added comments

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+	r.BuildCount.WithLabelValues(buildLabel).Inc()
+}
+
+func (r CustomBuildMetrics) BuildRunEstablishSet(buildLabel string, duration time.Duration) {
```
pls add comments to the function

zhangtbj replied:
added comments

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+	r.BuildRunEstablishDuration.WithLabelValues(buildLabel).Set(duration.Seconds())
+}
+
+func (r CustomBuildMetrics) BuildRunCountInc(buildLabel string) {
```
pls add comments to the function

zhangtbj replied:
added comments

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+	r.BuildRunCount.WithLabelValues(buildLabel).Inc()
+}
+
+func (r CustomBuildMetrics) BuildRunReadySet(buildLabel string, duration time.Duration) {
```
pls add comments to the function

zhangtbj replied:
added comments

qu1queee commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,129 @@
+	r.BuildRunReadyDuration.WithLabelValues(buildLabel).Set(duration.Seconds())
+}
+
+func (r CustomBuildMetrics) BuildRunReadyObserve(buildLabel string, duration time.Duration) {
```
pls add comments to the function

zhangtbj replied:
added comments

qu1queee commented:
I spoke with @zhangtbj on this, will be better pairing next week to finalize this

brancz commented on docs/metrics.md:
```diff
@@ -0,0 +1,14 @@
+| Name | Type | Description | Status |
+| ---- | ---- | ----------- | ------ |
+| `up_status` | Gauge | The up status of the build-controller. | experimental |
+| `build_count` | Gauge | Build creation count. | experimental |
```
This seems to be a counter? If so the type should be counter and the naming best practice is to use the `_total` suffix.

zhangtbj replied:
Hi @brancz ,

Yes, I considered that before, but in future the build_count maybe dynamically increase or decrease if end user create or delete a build.

That is the plan, so I keep to use Gauge here.

brancz replied:
Could you explain this in the help text? The convention then would be just `builds`, the `_count` suffix here is going to cause confusion much like I was :)

zhangtbj replied:
Yes, I agree. Let me change the build and buildrun count to counter type now. If we want to support complex case like increase or decrease, let us change them in future.

brancz replied:
Sounds good!

brancz replied:
One addition, if this is about creations, I would recommend to reflect that in the metric name as in: `builds_created_total`

zhangtbj replied:
Hi @brancz ,

Yes, I followed the official naming `<component_name>_<metric>` in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md to change them to:

- build_builds_total
- build_buildruns_total
- build_buildrun_establish_duration_seconds
- build_buildrun_completion_duration_seconds

in the lates commit

brancz replied:
as this describes the creation, I would recommend to include that in the metric name

zhangtbj replied:
Yep, @SaschaSchwarze0  also discussed in the below review comment.

Because it is not count from creation, I think it is better:
- build_builds_registered_total
- build_buildruns_completed_total

brancz commented on docs/metrics.md:
```diff
@@ -0,0 +1,14 @@
+| ---- | ---- | ----------- | ------ |
+| `up_status` | Gauge | The up status of the build-controller. | experimental |
+| `build_count` | Gauge | Build creation count. | experimental |
+| `buildrun_count` | Gauge | BuildRun execution count. | experimental |
```
ditto

zhangtbj replied:
Same as above ^^

zhangtbj replied:
Fixed all names as discussed, resolve the conversations. Thanks!

brancz commented on docs/metrics.md:
```diff
@@ -0,0 +1,14 @@
+
+| Name | Type | Description | Status |
+| ---- | ---- | ----------- | ------ |
+| `up_status` | Gauge | The up status of the build-controller. | experimental |
```
Prometheus has an automatic `up` metric that indicates whether a target can be scraped. This is typically used for "up" monitoring.

zhangtbj replied:
I didn't find this up metric, is it also related to our build controller `up` status? Or a general up status?

brancz replied:
I don't know how you set this metric, but generally speaking you want to avoid things like this. You want to instead expose raw data and make the decision of "up" or "down" based on the raw data and not create what is essentially a bespoke monitoring system in your component.

zhangtbj replied:
Hi @adambkaplan ,
How can we get the `version info` in the metric or the code logic?

zhangtbj replied:
I compared other community metric designs, there is no up_status.

So let me remove it in this initial PR first.

I removed in the latest commit: https://github.com/redhat-developer/build/pull/235/commits/248b3db1c98c9c36a87cab7f6f75f2095cb8400a

brancz commented on docs/metrics.md:
```diff
@@ -0,0 +1,14 @@
+| `buildrun_count` | Gauge | BuildRun execution count. | experimental |
+| `buildrun_establish_duration` | Gauge | BuildRun establish duration in seconds. | experimental |
+| `buildrun_ready_duration` | Gauge | BuildRun ready duration in seconds. | experimental |
+| `buildrun_ready_duration_his_[start, width, count]` | Histogram | BuildRun ready duration histogram in seconds. | experimental |
```
Why the histogram and gauge about this? What's the difference?

The naming best practice about this would be just the `_duration_seconds` suffix. And for me reading this, I don't actually understand what this metric tells me.

zhangtbj replied:
- buildrun_ready_duration: I used it for the trend to show how long recent build runs take
- buildrun_ready_duration_his: I used it for the statistics to classify different execution times to different buckets

<img width="566" alt="截屏2020-06-10 下午1 59 23" src="https://user-images.githubusercontent.com/18108042/84232325-a42ef500-ab22-11ea-8471-81d3e3806b16.png">

brancz replied:
These sound like they should be instrumented in the component that actually performs the build run.

zhangtbj replied:
What is your mean? They are all calculate during buildrun controller execution logic

brancz replied:
The controller doesn't actually execute the build. The component that does should expose that metric.

zhangtbj replied:
Yep, I also compared other community metric design, like tekton pipeline:
https://github.com/tektoncd/pipeline/blob/master/docs/metrics.md

The duration metrics are all `Histogram` types, so I also removed the `Gauge` one, just keep the `Histogram` one:
```
	buildRunEstablishDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name: "build_buildrun_establish_duration_seconds",
			Help: "BuildRun establish duration in seconds.",
			Buckets: prometheus.LinearBuckets(0, 0.5, 10),
		},
		[]string{"build_buildstrategy", "buildrun_namespace"})

	buildRunCompletionDuration = prometheus.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "build_buildrun_completion_duration_seconds",
			Help:    "BuildRun completion duration in seconds.",
			Buckets: prometheus.LinearBuckets(50, 50, 10),
		},
		[]string{"build_buildstrategy", "buildrun_namespace"})
```

The change is in the latest commit: https://github.com/redhat-developer/build/pull/235/commits/248b3db1c98c9c36a87cab7f6f75f2095cb8400a

brancz replied:
I'm a little confused. If tekton already has these metrics, why do we need them again?

zhangtbj replied:
No, Tekton just has the metrics about `taskrun`, they don't have our metrics like buildrun, I just compare the metric types with Tekton metrics.

brancz replied:
I see, I didn't realize they are distinct concepts.

brancz commented on docs/metrics.md:
```diff
@@ -0,0 +1,14 @@
+| `up_status` | Gauge | The up status of the build-controller. | experimental |
+| `build_count` | Gauge | Build creation count. | experimental |
+| `buildrun_count` | Gauge | BuildRun execution count. | experimental |
+| `buildrun_establish_duration` | Gauge | BuildRun establish duration in seconds. | experimental |
```
Can you explain in the help text what "established" means?

zhangtbj replied:
`buildRunEstablishDuring = buildRun.Status.StartTime.Time - buildRun.CreationTimestamp.Time`

It is the preparation time in buildrun controller logic

brancz replied:
That makes a lot of sense, can you reflect that in the help text? :)

sbose78 commented:
@zhangtbj @qu1queee ,

Could we ensure adherence to guidelines in 
* https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md
* https://prometheus.io/docs/practices/naming/

qu1queee commented:
@sbose78 definitely. Sorry I have been busy this week, but I will sync with @zhangtbj on this PR while I also reviewed long time ago.

zhangtbj commented:
Hi @sbose78 and @qu1queee and all,

I have followed the official naming to refine all metric names:
- build_builds_total (Counter)
- build_buildruns_total (Counter)
- build_buildrun_establish_duration_seconds (Histogram)
- build_buildrun_completion_duration_seconds (Histogram)

And also addressed all the comments. Please review again.

<img width="1338" alt="截屏2020-06-21 上午10 11 12" src="https://user-images.githubusercontent.com/18108042/85215477-f0442a00-b3ab-11ea-96d2-69e4a6387175.png">


Sorry, I didn't do the commit well, so there are some commits. Let us use this PR for reviewing.

If it is ok, I can provide another clear PR again.

This is the first initial metric enablement. I suggest we just keep the basic metrics and keep improving it in future PRs.

So please help review and move forward.

Thanks a lot for all of your comments and suggestions!

SaschaSchwarze0 commented on docs/metrics.md:
```diff
@@ -0,0 +1,12 @@
+
+| Name | Type | Description | Label | Status |
+| ---- | ---- | ----------- | ----- | ------ |
+| `build_builds_total` | Counter | Number of total created Builds. | build_buildstrategy=<build_buildstrategy_name> | experimental |
+| `build_buildruns_total` | Counter | Number of total new triggered BuildRuns. | build_buildstrategy=<build_buildstrategy_name> | experimental |
+| `build_buildrun_establish_duration_seconds` | Histogram | BuildRun establish duration in seconds. | build_buildstrategy=<build_buildstrategy_name><br>buildrun_namespace=<buildrun_namespace> | experimental |
+| `build_buildrun_completion_duration_seconds` | Histogram | BuildRun completion duration in seconds. | build_buildstrategy=<build_buildstrategy_name><br>buildrun_namespace=<buildrun_namespace> | experimental |
```
I suggest to have the namespace as label in all metrics. Also, I suggest shortening the labels to `namespace` and `buildstrategy`.

I'd like to see descriptions for the metrics that clearly describe what the metric measures. Currently only the last one fits the implementation:

- build_builds_total: the description is fine with me (I would remove the total out of it but that's only cosmetic). But, the implementation measures something completely different. It measures all creations and updates of a build resource that leads to a Build with Registered=True. That's something completely different. I comment this in the code again.
- build_buildruns_total: the description is unclear because the term _triggered_ is not defined in the context of this project. I suggest we measure what the name specifies = the number of created buildruns. The implementation currently measures the number of completed buildruns. I comment this in the code again.
- build_buildrun_establish_duration_seconds: the name and description are unclear because the term _established_ is not defined in the context of this project. I suggest to name it like `build_buildrun_created_to_running_duration_seconds` with a corresponding description `Duration in seconds of a build run from its creation to its start.`
- build_buildrun_completion_duration_seconds: the name and description are okay but not specific enough I think. I would name it `build_buildrun_created_to_completed_duration_seconds` with a description `Duration in seconds of a build run from its creation to its completion.`

zhangtbj replied:
Yes, I think about two questions when I doing the development:
- **I suggest to have the namespace as label in all metrics.**
I used namespace as label for all metrics, but at that time, I also face the metric limitation problem. I think it is not good like tekton to export lots of metrics which make the monitoring system crash. So I reduce the build and buildrun count just for buildstrategy. I also would like to add more label like namespace. But not sure if it will generated lots of metrics when we use both buildstrategy and namespace for all custom metrics.

- **Also, I suggest shortening the labels to namespace and buildstrategy.**
I also used the short name before, but I found there are some labels in the system which may make the end user confusing: such as: namespace, namespace.name or kubernetes.namespace.name. 
But I agree we can change to short name if we have a good document.
<img width="244" alt="截屏2020-06-22 下午3 53 39" src="https://user-images.githubusercontent.com/18108042/85262782-d8020700-b4a0-11ea-9ba1-a207c4a8370d.png">



About the name and description, how about:
- build_builds_registered
- build_buildruns_completed
- `build_buildrun_created_to_running_duration_seconds` is too long, I think we can have a short name but have a better description
- Same for `build_buildrun_created_to_completed_duration_seconds`

zhangtbj replied:
Changed to use:
- build_builds_registered_total
- build_buildruns_completed_total

And short the label to `buildstrategy` and `namespace`

SaschaSchwarze0 commented on pkg/controller/build/build_controller.go:
```diff
@@ -162,6 +163,10 @@ func (r *ReconcileBuild) Reconcile(request reconcile.Request) (reconcile.Result,
 		return reconcile.Result{}, err
 	}
+
+	// Increase Build count in metrics
+	buildmetrics.BuildCountInc(b.Spec.StrategyRef.Name)
```
This increments the metric for each successful reconcile of a build and not only on its creation. Failed registers are not counted. My gut feeling is that the creation predicate function is the only place where one can be 100 % sure that one currently handles a creation (see https://github.com/SaschaSchwarze0/build/blob/985d2716a0f38ce83885751333acae9c5273bc2f/pkg/controller/buildrun/buildrun_controller.go#L80-L83). I checked whether side effects in predicate functions are a bad practice but did not find something.

zhangtbj replied:
predicate is better to cover all succeed and fail case.

But for the build, there is also another problem here.

If the controller is restarted, the count will be clean to zero, and `predicate` way will start count the new created build or buildrun from zero.

But count in the reconciliation logic is different. After you start a new controller pod, ALL builds will go through the reconcile logic again. It mean, you can still get the total succeed build count after the pod restart.

So I think the best way like what tekton does, by using the `lister` to query the real total or running taskrun correctly, it won't missed anyone.

both current `succeed` and `predicate` ways cannot cover all cases, that is why I suggest we can have a basic metrics first then continue improving our logic or metric together.

SaschaSchwarze0 commented on pkg/controller/buildrun/buildrun_controller.go:
```diff
@@ -180,12 +181,24 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
+			// Add BuildRun establish and completion times in metrics
+			buildRunCompletionDuring := buildRun.Status.CompletionTime.Time.Sub(buildRun.CreationTimestamp.Time)
+			buildRunEstablishDuring := buildRun.Status.StartTime.Time.Sub(buildRun.CreationTimestamp.Time)
+			buildmetrics.BuildRunEstablishObserve(build.Spec.StrategyRef.Name, buildRun.Namespace, buildRunEstablishDuring)
```
You capture this metric on completion. You can capture this earlier by using condition `lastTaskRun.Status.StartTime != nil && buildRun.Status.StartTime == nil`.

zhangtbj replied:
Yes, I also think about that.
If we use `lastTaskRun.Status.StartTime != nil && buildRun.Status.StartTime == nil`
We can cover more, but still cannot cover all :(
- We cover the buildrun count of taskrun is executed successfully
- We cover the buildrun count of taskrun is executed fail
- But we still cannot cover the buildrun count when fail to create sa, add ownereference or any failure before generate taskrun correctly.

So how should we name the metric? `build_buildruns_taskrun_succeedandfail` ?
So that is why I also suggest to improve our logic with metric in future together.

SaschaSchwarze0 commented on pkg/controller/buildrun/buildrun_controller.go:
```diff
@@ -180,12 +181,24 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
+			buildRun.Status.CompletionTime = lastTaskRun.Status.CompletionTime
+
+			// Increase BuildRun count in metrics
+			buildmetrics.BuildRunCountInc(build.Spec.StrategyRef.Name)
```
Hereby, you are only counting build runs that are complete. Any build run that fails on creation of the task run will be missed. Same suggestion as for the build count: the predicate function imo works best.

zhangtbj replied:
Yep, like my comments above, if we all think `predicate` is the better way at current stage, let us move to that one, I also think it is a better way to cover both succeed and fail cases.

But we also need to think the about the case after the controller is restarted and metric is cleaned.

SaschaSchwarze0 commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,70 @@
+			Name: "build_builds_total",
+			Help: "Number of total created Builds.",
+		},
+		[]string{"build_buildstrategy"})
```
I suggest to use constants for the label keys as they are used for multiple metrics.

zhangtbj replied:
Good suggestion! If we still need this PR, I can change that with the naming improvement in your today's first review.

Thanks!

zhangtbj replied:
Refine by using:
```
const (
	buildStrategyLabel string = "buildstrategy"
	namespaceLabel string = "namespace"
)
```

gabemontero commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,70 @@
+			Name: "build_buildruns_total",
+			Help: "Number of total new triggered BuildRuns.",
+		},
+		[]string{"build_buildstrategy"})
```
just call this `strategy`

zhangtbj replied:
done in the latest commit

gabemontero replied:
label names look good

gabemontero commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,70 @@
+			Help: "BuildRun establish duration in seconds.",
+			Buckets: prometheus.LinearBuckets(0, 0.5, 10),
+		},
+		[]string{"build_buildstrategy", "buildrun_namespace"})
```
just use `strategy` and `namespace`

zhangtbj replied:
done in the latest commit

gabemontero commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,70 @@
+			Help:    "BuildRun completion duration in seconds.",
+			Buckets: prometheus.LinearBuckets(50, 50, 10),
+		},
+		[]string{"build_buildstrategy", "buildrun_namespace"})
```
`strategy` and `namespace`

zhangtbj replied:
done in the latest commit

zhangtbj commented:
We also follow the way from the prometheus document for `init`, like this:

```
Register the metric so that prometheus will know to export it.

Example

 func init() {
   prometheus.MustRegister(requestCounter)
   prometheus.MustRegister(requestLatencies)
   prometheus.MustRegister(requestLatenciesSummary)
 }
```


https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md

zhangtbj commented:
Hi all,

I should addressed all comments, please let me know if I still missed any suggestion.

Or any other suggestion?

Thanks!

qu1queee commented:
@zhangtbj  this https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md is for educational purposes, also the code it reference is obsolete(pls double check), but the way the k8s is doing now metrics seems to be [this](https://github.com/kubernetes/kubernetes/blob/7151131d79674d073e716063a03f8cbd67671e33/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L229), where the `init` is not longer there. I can give you two reason why the init is not desirable:
1. Uncertainty on when the init for this package will be called, it depends a lot on other packages that exist.
2. Having the `init` inside the package will make it harder for unit test, e.g. How I can test other metrics that are not properly configured if I cannot register them

zhangtbj commented:
Hi @qu1queee ,

I didn't find any document or example about the way what you are using.  What you are show should be the `k8s.io/component-base/metrics/legacyregistry`, I am not sure if it is the same metric as prometheus.

What I did is follow the documents what I learn, I can give more info:
- https://godoc.org/github.com/prometheus/client_golang/prometheus
- https://github.com/prometheus/client_golang/blob/master/examples/random/main.go#L63

If the `init` is the official way for prometheus, it can be called directly and internally (by shard registry I remembered, but I cannot find that doc) and if it is an official way, there must be a way to test it. I don't understand what is `other metrics that are not properly configured if I cannot register them`? If you don't register it, why you want to use or test it?

Of course, if you can provide an official document or example, we can change to that way.

zhangtbj commented:
I found that, I followed this way to registry this custom metrics in the global registry from `controller-runtime/pkg/metrics`.
So we don't need any additional code, global registry will help us config and use the port 8383 same as controller metric:

**Publishing Additional Metrics**
https://book.kubebuilder.io/reference/metrics.html#publishing-additional-metrics

If you wish to publish additional metrics from your controllers, this can be easily achieved by using the global registry from `controller-runtime/pkg/metrics`.

One way to achieve this is to declare your collectors as global variables and then register them using init().

For example:

```
import (
    "github.com/prometheus/client_golang/prometheus"
    "sigs.k8s.io/controller-runtime/pkg/metrics"
)

var (
    goobers = prometheus.NewCounter(
        prometheus.CounterOpts{
            Name: "goobers_total",
            Help: "Number of goobers proccessed",
        },
    )
    gooberFailures = prometheus.NewCounter(
        prometheus.CounterOpts{
            Name: "goober_failures_total",
            Help: "Number of failed goobers",
        },
    )
)

func init() {
    // Register custom metrics with the global prometheus registry
    metrics.Registry.MustRegister(goobers, gooberFailures)
}
```

qu1queee commented:
@zhangtbj thanks for the links. I think https://godoc.org/github.com/prometheus/client_golang/prometheus stills as stated there, a basic example. Also, https://book.kubebuilder.io/reference/metrics.html#publishing-additional-metrics is accurate, while having the `init()` there will work. Where I´m trying to persuade you is that this `init()` implementation is too basic for us, and that we need more control on when we start registering the collectors. I gave previously an explanation around `Uncertainty on when the init for this package will be called, it depends a lot on other packages that are imported at the same time as the metrics one.`

Regarding your question
>  I don't understand what is other metrics that are not properly configured if I cannot register them? If you don't register it, why you want to use or test it?

All I tried to say is that for unit-testing, I would like to be able to register N Collectors because of X spec I might require to write.

Regarding 

> What you are show should be the k8s.io/component-base/metrics/legacyregistry, I am not sure if it is the same metric as prometheus.

yes, is prometheus. See https://github.com/kubernetes/component-base/blob/master/metrics/legacyregistry/registry.go#L56 , https://github.com/kubernetes/component-base/blob/master/metrics/registry.go#L143 and then https://github.com/kubernetes/component-base/blob/master/metrics/registry.go#L100-L102

I´m approving your PR, I leave to you the decision around changing the init.

zhangtbj commented:
Thanks everyone and welcome to the 100+ comments PR.

Please let me know if you have any other suggestion. If no objection, I will continue merging this PR tomorrow.

Thanks all!

qu1queee commented:
@zhangtbj from my side only to squash commits :)

SaschaSchwarze0 commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,75 @@
+	buildRunCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "build_buildruns_completed_total",
+			Help: "Number of total new completed BuildRuns.",
```
remove `new`

zhangtbj replied:
fix that

SaschaSchwarze0 commented on docs/metrics.md:
```diff
@@ -0,0 +1,12 @@
+
+| Name | Type | Description | Label | Status |
+| ---- | ---- | ----------- | ----- | ------ |
+| `build_builds_registered_total` | Counter | Number of total created Builds. | buildstrategy=<build_buildstrategy_name> | experimental |
+| `build_buildruns_completed_total` | Counter | Number of total new triggered BuildRuns. | buildstrategy=<build_buildstrategy_name> | experimental |
```
Can you correct the descriptions to `Number of registered Builds.` and `Number of completed BuildRuns.`

zhangtbj replied:
fix that.

zhangtbj commented:
Thanks @qu1queee ,

I already squash to just one commit in this PR: https://github.com/redhat-developer/build/pull/235/commits

Hi all, all comments should be addressed now.

Please review. 

Thanks!

SaschaSchwarze0 commented:
/lgtm

zhangtbj commented:
Hi all,

There are two approvals in the issue and two ok in the slack channel.

Based on the requirement, I will merge this PR first for the first basic metric enablement.

Let us use another PR to improve or enhance that continually.

Thanks a lot all!

zhangtbj commented:
/approve
